File: app/src/main/java/com/erpdevelopment/vbvm/fragments/studies/StudiesFragment.java
package com.erpdevelopment.vbvm.fragments.studies;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import com.raizlabs.android.dbflow.runtime.OnTableChangedListener;
import com.raizlabs.android.dbflow.sql.language.SQLite;
import com.raizlabs.android.dbflow.structure.BaseModel;
import com.zhuinden.simplestack.BackstackDelegate;

import com.erpdevelopment.vbvm.R;
import com.erpdevelopment.vbvm.api.APIManager;
import com.erpdevelopment.vbvm.api.DatabaseManager;
import com.erpdevelopment.vbvm.application.MainActivity;
import com.erpdevelopment.vbvm.fragments.shared.AbstractFragment;
import com.erpdevelopment.vbvm.fragments.studies.study.StudyKey;
import com.erpdevelopment.vbvm.model.Category;
import com.erpdevelopment.vbvm.model.Category$$Parcelable;
import com.erpdevelopment.vbvm.model.Category_Table;
import com.erpdevelopment.vbvm.model.Study;
import com.erpdevelopment.vbvm.model.Study_Table;
import com.erpdevelopment.vbvm.util.ServiceLocator;
import com.erpdevelopment.vbvm.views.LoadingView;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * A simple {@link Fragment} subclass.
 * to handle interaction events.
 * Use the {@link StudiesFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class StudiesFragment extends AbstractFragment {

    private static final String TAG = "StudiesFragment";
    /**
     * The {@link android.support.v4.view.PagerAdapter} that will provide
     * fragments for each of the sections. We use a
     * {@link FragmentPagerAdapter} derivative, which will keep every
     * loaded fragment in memory. If this becomes too memory intensive, it
     * may be best to switch to a
     * {@link android.support.v4.app.FragmentStatePagerAdapter}.
     */
    private SectionsPagerAdapter mSectionsPagerAdapter;

    private static boolean categoriesCompleted = false;
    private static boolean studiesCompleted = false;
    private static Date lastRequestDate = null;

    private static Runnable activeRunner = null;

    /**
     * The {@link ViewPager} that will host the section contents.
     */
    @BindView(R.id.studiesContainer)
    ViewPager mViewPager;

    @BindView(R.id.studiesToolar)
    Toolbar toolbar;

    @BindView(R.id.studiesTabs)
    TabLayout tabLayout;

    @BindView(R.id.loading_view)
    LoadingView loadingView;

    public StudiesFragment() {
        // Required empty public constructor
    }

    @Override
    public boolean shouldBitmapUI() {
        return true;
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment StudiesView.
     */
    public static StudiesFragment newInstance() {
        StudiesFragment fragment = new StudiesFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
        }
    }

    private void configureCategoryPager() {
        List<Category> categories = SQLite.select().from(Category.class).orderBy(Category_Table.order, true).queryList();
        if (SQLite.selectCountOf().from(Study.class).count() > 0) {
            mSectionsPagerAdapter.setCategories(categories);
            mViewPager.setVisibility(View.VISIBLE);
            loadingView.setVisibility(View.GONE);
            Log.d(TAG, "configureCategoryPager: " + this);
        }
    }

    private boolean hasContent() {
        if (SQLite.selectCountOf().from(Study.class).count() > 0 && SQLite.selectCountOf().from(Category.class).count() > 0)
            return true;
        return false;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_studies_root, container, false);
        unbinder = ButterKnife.bind(this, view);

        toolbar.setTitle(R.string.title_studies);

        mSectionsPagerAdapter = new SectionsPagerAdapter(this.getChildFragmentManager());
        mViewPager.setAdapter(mSectionsPagerAdapter);
        configureCategoryPager();

        tabLayout.setupWithViewPager(mViewPager);

        if (!hasContent()) {
            loadingView.setVisibility(View.VISIBLE);
            mViewPager.setVisibility(View.GONE);
        }

        if(lastRequestDate == null || TimeUnit.MILLISECONDS.toSeconds((new Date()).getTime() - lastRequestDate.getTime()) > 300 ) {
            categoriesCompleted = false;
            studiesCompleted = false;
            lastRequestDate = new Date();
            Log.d(TAG, "onCreateView: ");
            APIManager.getInstance().downloadCategories(success -> {
                // Categories downloaded... refresh that screen!
                if (success) {
                    //configureCategoryPager();
                    categoriesCompleted = true;
                    if (studiesCompleted && categoriesCompleted && StudiesFragment.activeRunner != null) {
                        StudiesFragment.activeRunner.run();
                    }
                }
            });

            APIManager.getInstance().downloadStudies(success -> {
                // Studies downloaded... refresh!
                if (success) {
                    studiesCompleted = true;
                    if (studiesCompleted && categoriesCompleted && StudiesFragment.activeRunner != null) {
                        StudiesFragment.activeRunner.run();
                    }
                }
            });
        }

        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        Log.d(TAG, "onAttach: " + this);
        activeRunner = new Runnable() {
            @Override
            public void run() {
                if (!isDetached())
                    configureCategoryPager();
            }
        };
    }



    @Override
    public void onDetach() {
        super.onDetach();
        mSectionsPagerAdapter = null;
        Log.d(TAG, "onDetach: " + this);
        activeRunner = null;
    }


    /**
     * A placeholder fragment containing a simple view.
     */
    public static class StudiesListFragment extends Fragment {
        /**
         * The fragment argument representing the section number for this
         * fragment.
         */
        private static final String ARG_SECTION_NUMBER = "section_number";
        private static final String ARG_CATEGORY = "ARG_CATEGORY";
        private Category category;

        private RecyclerView view;

        BroadcastReceiver studiesReceiver;

        public StudiesListFragment() {
        }

        /**
         * Returns a new instance of this fragment for the given section
         * number.
         */
        public static StudiesListFragment newInstance(Category category) {
            StudiesListFragment fragment = new StudiesListFragment();
            Bundle args = new Bundle();
            args.putParcelable(ARG_CATEGORY, new Category$$Parcelable(category));
            fragment.setArguments(args);
            return fragment;
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                 Bundle savedInstanceState) {

            View rootView = inflater.inflate(R.layout.fragment_studies, container, false);

            category = ((Category$$Parcelable)getArguments().getParcelable(ARG_CATEGORY)).getParcel();

            final List<Study> studies = SQLite.select().from(Study.class).where(Study_Table.category.eq(category.id)).orderBy(Study_Table.bibleIndex, true).queryList();
            DisplayMetrics displayMetrics = new DisplayMetrics();
            ((WindowManager)getContext().getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay().getMetrics(displayMetrics);

            Log.d("Placeholder", "onCreateView: found " + studies.size() + " studies");

            float dpWidth = displayMetrics.widthPixels / displayMetrics.density;

            float imageWidth = Math.max((dpWidth / 4) - 1, 120);

            float dpHeight = displayMetrics.heightPixels / displayMetrics.density;

            float imageHeight = dpHeight / 3;
            imageWidth = Math.min(imageHeight, imageWidth);

            int numberOfColumns = (int)(dpWidth / imageWidth);

            // on a large device go with 4 columns?

            if (rootView instanceof RecyclerView) {
                Context context = rootView.getContext();
                RecyclerView recyclerView = (RecyclerView) rootView;
                this.view = recyclerView;
                recyclerView.setLayoutManager(new GridLayoutManager(context, numberOfColumns));
                recyclerView.setAdapter(new MyStudyRecyclerViewAdapter(studies, numberOfColumns, new MyStudyRecyclerViewAdapter.OnStudyInteractionListener() {
                    @Override
                    public void studyClicked(Study study) {
                        APIManager.getInstance().downloadLessons(study.id, success -> {
                            Log.d(TAG, "Downloaded the lessons successfully " + success);
                        });
                        ((BackstackDelegate)ServiceLocator.getService(getContext(), MainActivity.StackType.STUDIES.name())).getBackstack().goTo(StudyKey.create(study.id));
                    }
                }));
            }

            final Handler mainHandler = new Handler(getContext().getMainLooper());

            studiesReceiver = new BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    mainHandler.post(() -> reloadData());
                }
            };

            LocalBroadcastManager.getInstance(getContext()).registerReceiver(studiesReceiver, new IntentFilter(Study.updated()));

            //DatabaseManager.observer.addOnTableChangedListener(tableChangedListener);
            return rootView;
        }

        public void reloadData() {
            MyStudyRecyclerViewAdapter adapter = (MyStudyRecyclerViewAdapter)view.getAdapter();
            final List<Study> studies = SQLite.select().from(Study.class).where(Study_Table.category.eq(category.id)).orderBy(Study_Table.bibleIndex, true).queryList();
            adapter.setStudies(studies);
        }

        @Override
        public void onDetach() {
            super.onDetach();
            //DatabaseManager.observer.removeTableChangedListener(tableChangedListener);
            Log.d("STUDIESFRAGMENT", "removed table change listener");
            if (studiesReceiver != null) {
                LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(studiesReceiver);
                studiesReceiver = null;
            }
        }
    }

    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    private class SectionsPagerAdapter extends FragmentPagerAdapter {
        private List<Category> categories = new ArrayList<>();

        public SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        public void setCategories(List<Category> categories) {
            if (categories == null) {
                categories = new ArrayList<>();
                notifyDataSetChanged();
                return;
            }
            this.categories = categories;
            notifyDataSetChanged();
        }

        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).
            return StudiesListFragment.newInstance(categories.get(position));
        }

        @Override
        public int getCount() {
            return categories.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            Category category = categories.get(position);
            return category.name;
        }
    }

}

File: app/src/main/java/com/erpdevelopment/vbvm/fragments/studies/StudiesFragment.kt
package com.erpdevelopment.vbvm.fragments.studies

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.os.Parcelable
import android.util.DisplayMetrics
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import com.erpdevelopment.vbvm.R
import com.erpdevelopment.vbvm.api.APIManager
import com.erpdevelopment.vbvm.application.MainActivity
import com.erpdevelopment.vbvm.fragments.shared.AbstractFragment
import com.erpdevelopment.vbvm.fragments.studies.study.StudyKey
import com.erpdevelopment.vbvm.model.StudyViewModel
import com.erpdevelopment.vbvm.util.ServiceLocator
import com.erpdevelopment.vbvm.views.LoadingView
import com.google.android.material.tabs.TabLayout
import com.raizlabs.android.dbflow.sql.language.SQLite
import com.uber.autodispose.android.lifecycle.autoDispose
import com.zhuinden.simplestack.BackstackDelegate
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.versebyverseministry.models.Category
import org.versebyverseministry.models.Study
import org.versebyverseministry.models.Study_Table
import org.versebyverseministry.models.`Category$$Parcelable`
import com.uber.autodispose.android.lifecycle.autoDispose


/**
 * A simple [Fragment] subclass.
 * to handle interaction events.
 * Use the [StudiesFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class StudiesFragment : AbstractFragment() {
    /**
     * The [PagerAdapter] that will provide
     * fragments for each of the sections. We use a
     * [FragmentPagerAdapter] derivative, which will keep every
     * loaded fragment in memory. If this becomes too memory intensive, it
     * may be best to switch to a
     * [FragmentStatePagerAdapter].
     */
    private var mSectionsPagerAdapter: SectionsPagerAdapter? = null
    /**
     * The [ViewPager] that will host the section contents.
     */
    private lateinit var mViewPager: ViewPager
    private lateinit var loadingView: LoadingView

    private lateinit var viewModel: StudiesViewModel
    private lateinit var studiesViewModel: StudyViewModel

    override fun shouldBitmapUI(): Boolean {
        return true
    }

    private fun configureCategoryPager() {
        if (Study.countOf() > 0) {
            mViewPager.visibility = View.VISIBLE
            loadingView.visibility = View.GONE
            Log.d(TAG, "configureCategoryPager: $this")
        }
    }

    private fun hasContent(): Boolean {
        return Study.countOf() > 0 && Category.countOf() > 0
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        Log.d(TAG, "onAttach: $this")
        viewModel = ViewModelProvider(this).get(StudiesViewModel::class.java)
        studiesViewModel = ViewModelProvider(this).get(StudyViewModel::class.java)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? { // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_studies_root, container, false)
        val toolbar: Toolbar = view.findViewById(R.id.studiesToolar)
        val tabLayout: TabLayout = view.findViewById(R.id.studiesTabs)
        loadingView = view.findViewById(R.id.loading_view)
        mViewPager = view.findViewById(R.id.studiesContainer)

        toolbar.setTitle(R.string.title_studies)
        mSectionsPagerAdapter = SectionsPagerAdapter(viewModel.categories, this.childFragmentManager)
        mViewPager.setAdapter(mSectionsPagerAdapter)
        configureCategoryPager()
        tabLayout.setupWithViewPager(mViewPager)
        if (!hasContent()) {
            loadingView.setVisibility(View.VISIBLE)
            mViewPager.setVisibility(View.GONE)
        }

        return view
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        studiesViewModel.bibleOrderedStudies.observe(viewLifecycleOwner, Observer { studies ->
            Log.d(TAG, "🙌 There are ${studies.size} studies in the room database and we observed them!!")
        })

        viewModel.isAllDataLoaded.observe(viewLifecycleOwner, androidx.lifecycle.Observer { loaded ->
            if (loaded) {
                configureCategoryPager()
            }
        })

        viewModel.viewEvents.observeOn(AndroidSchedulers.mainThread())
                .autoDispose(this, Lifecycle.Event.ON_DESTROY)
                .subscribe {
                    when (it) {
                        is CategoriesUpdated -> {
                            mSectionsPagerAdapter?.categoriesUpdated()
                        }
                        else -> {
                            Log.d(TAG, "An event happened: ${it}")
                        }
                    }
                }
    }

    override fun onDetach() {
        super.onDetach()
        mSectionsPagerAdapter = null
        Log.d(TAG, "onDetach: $this")
    }

    /**
     * A placeholder fragment containing a simple view.
     */
    class StudiesListFragment : Fragment() {
        private var category: org.versebyverseministry.models.Category? = null
        private var view: RecyclerView? = null
        private var studiesReceiver: BroadcastReceiver? = null

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                                  savedInstanceState: Bundle?): View? {
            val rootView = inflater.inflate(R.layout.fragment_studies, container, false)
            val category = (arguments!!.getParcelable<Parcelable>(ARG_CATEGORY) as `Category$$Parcelable`).parcel
            this.category = category
            val studies = Study.fetchStudiesWithCategory(category)
            val displayMetrics = DisplayMetrics()
            (context!!.getSystemService(Context.WINDOW_SERVICE) as WindowManager).defaultDisplay.getMetrics(displayMetrics)
            Log.d("Placeholder", "onCreateView: found " + studies.size + " studies")
            val dpWidth = displayMetrics.widthPixels / displayMetrics.density
            var imageWidth = Math.max(dpWidth / 4 - 1, 120f)
            val dpHeight = displayMetrics.heightPixels / displayMetrics.density
            val imageHeight = dpHeight / 3
            imageWidth = Math.min(imageHeight, imageWidth)
            val numberOfColumns = (dpWidth / imageWidth).toInt()
            // on a large device go with 4 columns?
            if (rootView is RecyclerView) {
                val context = rootView.getContext()
                val recyclerView = rootView
                view = recyclerView
                recyclerView.layoutManager = GridLayoutManager(context, numberOfColumns)
                recyclerView.adapter = MyStudyRecyclerViewAdapter(studies, numberOfColumns, MyStudyRecyclerViewAdapter.OnStudyInteractionListener { study ->
                    APIManager.getInstance().downloadLessons(study.id, { success: Boolean -> Log.d(TAG, "Downloaded the lessons successfully $success") })
                    (ServiceLocator.getService<Any>(getContext(), MainActivity.StackType.STUDIES.name) as BackstackDelegate).backstack.goTo(StudyKey.create(study.id))
                })
            }
            val mainHandler = Handler(context!!.mainLooper)
            val receiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    mainHandler.post { reloadData() }
                }
            }
            studiesReceiver = receiver
            LocalBroadcastManager.getInstance(context!!).registerReceiver(receiver, IntentFilter(org.versebyverseministry.models.Study.updated()))
            //DatabaseManager.observer.addOnTableChangedListener(tableChangedListener);
            return rootView
        }

        fun reloadData() {
            val adapter = view!!.adapter as MyStudyRecyclerViewAdapter?
            val studies = SQLite.select().from(org.versebyverseministry.models.Study::class.java).where(Study_Table.category.eq(category!!.id)).orderBy(Study_Table.bibleIndex, true).queryList()
            adapter!!.setStudies(studies)
        }

        override fun onDetach() {
            super.onDetach()
            //DatabaseManager.observer.removeTableChangedListener(tableChangedListener);
            Log.d("STUDIESFRAGMENT", "removed table change listener")
            studiesReceiver?.let {
                LocalBroadcastManager.getInstance(context!!).unregisterReceiver(it)
                studiesReceiver = null
            }
        }

        companion object {
            /**
             * The fragment argument representing the section number for this
             * fragment.
             */
            private const val ARG_SECTION_NUMBER = "section_number"
            private const val ARG_CATEGORY = "ARG_CATEGORY"
            /**
             * Returns a new instance of this fragment for the given section
             * number.
             */
            fun newInstance(category: org.versebyverseministry.models.Category?): StudiesListFragment {
                val fragment = StudiesListFragment()
                val args = Bundle()
                args.putParcelable(ARG_CATEGORY, `Category$$Parcelable`(category))
                fragment.arguments = args
                return fragment
            }
        }
    }

    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    private inner class SectionsPagerAdapter(val categories: List<Category>, fm: FragmentManager?) : FragmentPagerAdapter(fm!!) {

        fun categoriesUpdated() {
            notifyDataSetChanged()
        }

        override fun getItem(position: Int): Fragment {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).
            return StudiesListFragment.newInstance(categories[position])
        }

        override fun getCount(): Int {
            return categories.size
        }

        override fun getPageTitle(position: Int): CharSequence? {
            val category = categories[position]
            return category.name
        }
    }

    companion object {
        private const val TAG = "StudiesFragment"
        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment StudiesView.
         */
        @JvmStatic
        fun newInstance(): StudiesFragment {
            val fragment = StudiesFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}